### ISSUE TITLE
Allow more than 1 TypeByte

### ISSUE THREAD

**Issue by ethanfrey** [tendermint/go-amino#21]
I am sure 256 types seems like quite a few implementations, but slowly this is becoming an issue with basecoin aka quark

We define one Tx struct and a TxInner interface that all transactions must implement in every app.  This allows us to very easily marshal and unmarshal both json and binary data, handle nested tx transparently, etc. I am very happy with this.  However.... we currently have 15 tx types in the core modules, and maybe 5 more in planning... plus at least 10 more in implementing elections.  And trying to avoid accidental collisions, rigel and I have been spacing these out a bit.  No real problem so far, but as soon as third-parties start writing modules and defining their own tx types (trackomatron, gaia, basecoin-stake, etc.) the number will get larger and will need to coordinate between multiple authors to avoid collisions. Or have modules that cannot both be installed in the same app due to this issue.

I would like to have 2 type bytes for tx, use the first 12 bits for the modules and the last 4 (last hex digit) for the tx in the repo itself. Maybe even consider 3 bytes... but maybe that is too much.  Anyway, flexibility in our encoding system would be great.

I understand that go-wire is used in many different places in tendermint core, where this limit is not a problem, and they wish to keep 1 byte prefixes, and we must maintain the same binary encoding.

I would propose a system where `wire.RegisterInterface` maintains the exact same semantics along with the same 1 byte prefix, but there is another function, eg. `wire.RegisterLongerInterface(interface{}, prefix int, ...LongerConcreteTypes )` and `LongerConcreteTypes` allows []byte instead of byte, and asserts that all types have the same length as defined in the call.

This can create and enhanced `TypeInfo`, that knows how many bytes it should take off for the comparison (as it is registered per interface), and extends `ByteToType` and `TypeToByte`.  I would also expand `data.Mapper.RegisterImplementation` to accept multiple bytes and do the same assertions (all same prefix length)  internally.

In short: `func (m Mapper) RegisterImplementation(data interface{}, kind string, b byte) Mapper` would become `func (m Mapper) RegisterImplementation(data interface{}, kind string, b ...byte) Mapper` and drop in compatible with any code that uses it.

If all modules now use this mapper instead of directly calling `wire.RegisterInterface`, I could change the functionality there directly instead of implementing `RegisterLongerInterface`

**Comment by rigelrozanski** [comment]
Just noticed that it doesn't appear that `kind` within `RegisterImplementation` is used? There to fulfill and interface? I agree that it makes sense to allow for more flexibility by extending the typebytes array. Allowing for layers of typebyte would allow for easier extensibility - aka each zone in a COSMOS hub could have its own typebyte space. It's a bit more clear if we can stick to using the `wire.RegisterInterface` - We should be able to have a single TB within your proposed `LongerConcreteTypes` so why not just extend what's already there instead of introducing a whole new interface?

**Comment by ethanfrey** [comment]
It's used by the json mapper

Please look at the `data` subdir... not too much and that is the interface I've been porting all code to use.

The problem with layers of type bytes is the json format...

```
{
  "type": "my-company",
  "data": {
    "type": "my-repo",
    "data": {
      "type": "my-tx",
       "data": ...stuff here...
    }
   }
}
```

And that at every level of wrapping... I think the users of the API would kill us when the json is 17 layers deep.

**Comment by ethanfrey** [comment]
Please look at the `data` subdir... not too much and that is the interface I've been porting all code to use.

**Comment by ebuchman** [comment]
Is it possible to support sub-types in go-wire yet ? If not, would this be a better approach? It would basically amount to register another interface as a concrete implementation of a type, so you'd need more than one type byte to resolve to a real concrete type

**Comment by liamsi** [comment]
Just cleaning up some open issues: This does not hold true anymore:
https://github.com/tendermint/go-amino#prefix-bytes-to-identify-the-concrete-type
